lib/segment/src/index/hnsw_index/graph_links/serializer.rs

```diff
@@ -1,356 +1,214 @@
     }
 
-    pub fn to_graph_links_ram(&self) -> GraphLinks {
-        let format = match &self.kind {
-            Kind::Uncompressed { .. } => GraphLinksFormat::Plain,
-            Kind::Compressed { .. } => GraphLinksFormat::Compressed,
-            Kind::CompressedWithVectors { .. } => GraphLinksFormat::CompressedWithVectors,
+    // 5. Write neighbors (and calculate `offsets`)
+    let mut links_buf = Vec::new();
+    let mut offset = 0; // elements for Plain, bytes for Compressed/CompressedWithVectors
+    let mut offsets = Vec::with_capacity(total_offsets_len as usize);
+    offsets.push(0);
+    for level in 0..levels_count {
+        let count = point_count_by_level.iter().skip(level).sum::<u64>() as usize;
+        let (level_m, mut iter) = match level {
+            0 => (hnsw_m.m0, Either::Left(0..count)),
+            _ => (hnsw_m.m, Either::Right(back_index[..count].iter().copied())),
         };
 
-        let size = self.level_offsets.as_bytes().len()
-            + self.reindex.as_bytes().len()
-            + self.neighbors.len()
-            + (match &self.kind {
-                Kind::Uncompressed {
-                    offsets_padding,
-                    offsets,
-                } => size_of::<HeaderPlain>() + offsets_padding + offsets.as_bytes().len(),
-                Kind::Compressed {
-                    compressed_offsets,
-                    offsets_parameters: _,
-                } => size_of::<HeaderCompressed>() + compressed_offsets.len(),
-                Kind::CompressedWithVectors {
-                    neighbors_padding,
-                    compressed_offsets,
-                    offsets_parameters: _,
-                    vector_layout: _,
-                } => {
-                    size_of::<HeaderCompressedWithVectors>()
-                        + neighbors_padding
-                        + compressed_offsets.len()
+        iter.try_for_each(|id| {
+            let mut raw_links = std::mem::take(&mut edges[id][level]);
+            match format_param {
+                GraphLinksFormatParam::Plain => {
+                    writer.write_all(raw_links.as_bytes())?;
+                    offset += raw_links.len();
+                }
+                GraphLinksFormatParam::Compressed => {
+                    pack_links(&mut links_buf, &mut raw_links, bits_per_unsorted, level_m);
+                    writer.write_all(&links_buf)?;
+                    offset += links_buf.len();
+                }
+                GraphLinksFormatParam::CompressedWithVectors(vectors) => {
+                    // Unwrap safety: `vector_layout` is `Some` for `CompressedWithVectors`.
+                    let vector_layout = vector_layout.unwrap();
+
+                    // 1. The varint-encoded length (`N` in the doc).
+                    writer.write_varint(raw_links.len() as u64)?;
+                    offset += u64::required_space(raw_links.len() as u64);
```

Comment:
I don't see this logic
```
offset += u64::required_space(raw_links.len() as u64);
```
In dev, can you please explain why you use `required_space` here instead of direct `raw_links.len()`? I expect you collect all writed bytes count into writer.